source/common/config/grpc_mux_impl.cc

```diff
@@ -116,17 +116,30 @@ ScopedResume GrpcMuxImpl::pause(const std::vector<std::string> type_urls) {
 void GrpcMuxImpl::onDiscoveryResponse(
     std::unique_ptr<envoy::service::discovery::v3::DiscoveryResponse>&& message,
     ControlPlaneStats& control_plane_stats) {
-  const std::string& type_url = message->type_url();
+  std::string type_url = message->type_url();
   ENVOY_LOG(debug, "Received gRPC message for {} at version {}", type_url, message->version_info());
   if (message->has_control_plane()) {
     control_plane_stats.identifier_.set(message->control_plane().identifier());
   }
-  if (api_state_.count(type_url) == 0) {
-    ENVOY_LOG(warn, "Ignoring the message for type URL {} as it has no current subscribers.",
-              type_url);
-    // TODO(yuval-k): This should never happen. consider dropping the stream as this is a
-    // protocol violation
-    return;
+  // If this type url is not watched(no subscriber or no watcher), try older version of type url.
+  if (api_state_.count(type_url) == 0 ||
+      (api_state_[type_url].watches_.empty() && !message->resources().empty())) {
+    absl::optional<std::string> old_type_url =
+        ApiTypeOracle::getEarlierTypeUrl(message->type_url());
```

Comment:
nit: pass `type_url` as a parameter to keep consistency in this scope

Reply:
Thanks, will do.